packages/libp2p/test/connection-manager/auto-dial.spec.ts

```diff
@@ -224,4 +226,56 @@ describe('auto-dial', () => {
     // should only have queried peer store once
     expect(peerStoreAllSpy.callCount).to.equal(1)
   })
+
+  it('should not re-dial peers we have recently failed to dial', async () => {
+    const peerWithAddress: Peer = {
+      id: await createEd25519PeerId(),
+      protocols: [],
+      addresses: [{
+        multiaddr: multiaddr('/ip4/127.0.0.1/tcp/4001'),
+        isCertified: true
+      }],
+      metadata: new Map(),
+      tags: new Map()
+    }
+    const undialablePeer: Peer = {
+      id: await createEd25519PeerId(),
+      protocols: [],
+      addresses: [{
+        multiaddr: multiaddr('/ip4/127.0.0.1/tcp/4002'),
+        isCertified: true
+      }],
+      // we failed to dial them recently
+      metadata: new Map([[LAST_DIAL_FAILURE_KEY, uint8ArrayFromString(`${Date.now() - 10}`)]]),
+      tags: new Map()
+    }
+
+    await peerStore.save(peerWithAddress.id, peerWithAddress)
+    await peerStore.save(undialablePeer.id, undialablePeer)
+
+    const connectionManager = stubInterface<ConnectionManager>({
+      getConnectionsMap: new PeerMap(),
+      getDialQueue: []
+    })
+
+    autoDialler = new AutoDial({
+      peerStore,
+      connectionManager,
+      events
+    }, {
+      minConnections: 10,
+      autoDialInterval: 10000
+    })
+    autoDialler.start()
+    void autoDialler.autoDial()
```

Comment:
See my other comment as well, but do we need all the test scaffolding for auto-dialer or can we just attempt to dial the peer directly?

Reply:
regardless of the answer here, we may want another test regarding direct-dials covering the expected cases (dial them again even if failed auto dial previously, because it's a direct dial request)

Reply:
The last-failed-dial key is only used by the auto-dialer to exclude those peers from autodialing, it's not used by explicit dials.

Reply:
I see that now, thanks